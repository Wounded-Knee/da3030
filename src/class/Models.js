const MODEL_TYPES = {
	GENERIC: 'Generic',
	CLOWN: 'Clown',
};
const EVENT_TYPES = {
	CREATED: 'Created',
	SERIALIZED: 'Serialized',
	DESERIALIZED: 'Deserialized',
};

class Generic {
	constructor(data) {
		this.data = data;
		this.modelName = MODEL_TYPES.GENERIC;

		this.data = {
			...data,
			meta: {
				...this.data.meta,
				modelName: this.modelName,
				events: [
					...this.data.meta.events || [],
					{
						userId: 0,
						date: new Date(),
						type: EVENT_TYPES.CREATED,
					}
				],
			}
		};
	}

	get(attribute) {
		return attribute ? this.data[attribute] : this.data;
	}

	serialize() {
		return {
			...this.data,
			meta: {
				...this.data.meta,
				events: [
					...this.data.meta.events,
					{
						userId: 0,
						date: new Date(),
						type: EVENT_TYPES.SERIALIZED,
					}
				],
			}
		};
	}

	deserialize(data) {
		this.data = {
			...data,
			meta: {
				...data.meta,
				events: [
					...data.meta.events,
					{
						userId: 0,
						date: new Date(),
						type: EVENT_TYPES.DESERIALIZED,
					}
				],
			}
		};
		return this.data;
	}
}

class Clown extends Generic {
	constructor(data) {
		const rv = super(data);
		this.modelName = MODEL_TYPES.CLOWN;
		return rv;
	}
}

export {
	MODEL_TYPES,
	EVENT_TYPES,
	Generic,
	Clown,
}