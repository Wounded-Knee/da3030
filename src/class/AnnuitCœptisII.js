import * as Models from './Models';
import { LOCAL_STORAGE_NAMES } from '../PROPHET60091';
const {
	MODEL_TYPES,
	ATTRIBUTE_NAMES,
	User
} = Models;
var AnnuitCœptisIIData = [];

class AnnuitCœptisII {
	constructor({ loadFile, onChange } = {}) {
		this.iterator = 0;
		this.dirty = false;
		this.onChange = onChange || function() {};
		this.load(loadFile);
		this.models = Models;
	}

	somethingChanged(model = undefined) {
		this.onChange();
	}

	create(data, modelType = MODEL_TYPES.GENERIC) {
		const model = Models[modelType];
		if (model) {
			const newNode = new model({
				...data,
				[ATTRIBUTE_NAMES.META]: {
					[ATTRIBUTE_NAMES.ID]: this.iterator++,
					[ATTRIBUTE_NAMES.MODEL_TYPE]: modelType,
				},
			}, this);
			AnnuitCœptisIIData.push(newNode);
			this.dirty = true;
			return newNode;
		} else {
			console.log(Models);
			throw new Error(`${modelType} does not exist.`);
		}
	}

	filter(callback) {
		return AnnuitCœptisIIData.filter(callback);
	}

	map(callback) {
		return AnnuitCœptisIIData.map(callback);
	}

	isDirty() {
		return this.dirty || this.map(
			node => node.isDirty()
		).reduce(
			(accumulator, value) => accumulator || value,
			this.dirty,
		);
	}

	save(file = 'AnnuitCœptisII') {
		const serializedData = this.map(
			node => node.save()
		);
		localStorage.setItem(file, JSON.stringify(serializedData));
		this.dirty = false;
	}

	load(file = 'AnnuitCœptisII') {
		AnnuitCœptisIIData = (JSON.parse(localStorage.getItem(file)) || []).map(
			nodeData => {
				const modelType = nodeData[ATTRIBUTE_NAMES.META][ATTRIBUTE_NAMES.MODEL_TYPE];
				const model = Models[modelType];
				return new model(nodeData, this);
			}
		);
		this.dirty = false;
		this.iterator = this._getFreshId();
	}

	// Force flag, when true, will clear even if dirty.
	// Otherwise, dirty flag acts as a safety.
	clear(force = false) {
		if (this.isDirty() && !force) {
			return false;
		}
		AnnuitCœptisIIData.length = 0;
		this.dirty = false;
		return true;
	}

	getById(id) {
		return this.filter(
			node => node.getId() === id
		)[0];
	}

	getData() {
		return AnnuitCœptisIIData;
	}

	getNotifications(notificationType) {
		return [];
	}

	getCurrentUser() {
		return (this.getByModelType(MODEL_TYPES.USER).filter(
			user => user.isCurrent()
		)[0]) || User.getAnonymous();
	}

	getTrailheads() {
		return this.getByModelType(MODEL_TYPES.TEXT_NODE).filter(
			textNode => !textNode.getParent()
		);
	}

	getByModelType(modelType) {
		return this.filter(
			node => node.getModelType() === modelType
		);
	}

	getLocalStorageInfo() {
		return {
			storageData: this.getLocalStorage(LOCAL_STORAGE_NAMES.DATA)
		};
	}

	getLocalStorage(fileName) {
		return JSON.parse(localStorage.getItem(fileName) || '{}');
	}

	setLocalStorage(fileName, data) {
		localStorage.setItem(
			fileName,
			JSON.stringify(data),
		);
		const newValue = this.getLocalStorage(fileName);
		console.log('setLocalStorage() ', newValue);
		return newValue;
	}

	represent() {
		return this.map(model => model.represent());
	}

	_getFreshId(nodes = AnnuitCœptisIIData) {
		return Math.max(
			Math.max.apply(
				Math,
				nodes.map(
					node => parseInt(node.id) || 0
				)
			) + 1
		, 0);
	}
}

export {
	AnnuitCœptisII as default,
	LOCAL_STORAGE_NAMES,
};