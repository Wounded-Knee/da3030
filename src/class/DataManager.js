class DataManager {
	constructor(config) {
		this.config = config;
		this.annuitCœptis = config.annuitCœptis;
		this.nodeType = config.nodeType;
	}

	getAll() {
		return this.filter( node => node.data.type === this.nodeType );
	}

	delete(node) {
		return this.annuitCœptis.remove(node);
	}

	update(node, data) {
		return this.annuitCœptis.update(node._id, data);
	}

	move(node, newParentNode) {
		const newNode = this._create(
			node.data,
			newParentNode,
		);
		this.delete(node);
		return newNode;
	}

	find(nodeId) {
		return this.annuitCœptis.find(nodeId);
	}

	getParentOf(childNode) {
		return this.filter(
			node => node.children.find(
				child => child._id === childNode._id
			)
		)[0];
	}

	getById(id) {
		return this.getAll().find(
			node => node.data.id === parseInt(id)
		);
	}

	filter(callback, startingPoint) {
		const tree =
			startingPoint ||
			this.annuitCœptis.getTree().data.filter(
				node => node.data.type === this.nodeType
			);

		return tree.length ? [
			...tree.filter(callback),
			...this.filter(
				callback,
				tree.reduce(
					(arr, val) => [ ...arr, ...val.children], []
				)
			)
		] : [];
	}

	create(data, parentNode = null) {
		return this._create(
			data,
			parentNode,
		);
	}

	_create(data, parentNode = null) {
		const newData = this._createNodeData(data);

		return newData ? this.annuitCœptis.addNode(
			this.annuitCœptis.createNode(
				parentNode,
				newData,
				{ annuitCœptis: 'annuitCœptis'}
			),
			parentNode
		) : false;
	}

	_createNodeData(nodeData) {
		const newId = this._getFreshId();
		const { userId } = this.annuitCœptis.getSettings();
		const newData = {
			...nodeData,
			id: newId,
			type: this.nodeType,
			authorId: userId,
		};

		return (
			nodeData instanceof Object
				? newData : false
		);
	}

	// Returns an unused ID
	// Finds the max ID int in the passed-in arguments, adds +1 to it.
	_getFreshId(nodes = this.getAll()) {
		return Math.max(
			Math.max.apply(
				Math,
				nodes.map(
					node => parseInt(node.data.id) || 0
				)
			) + 1
		, 0);
	}
};

export default DataManager;