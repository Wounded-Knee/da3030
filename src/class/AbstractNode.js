import DataManager from './DataManager';
const NODE_TYPES = {
  NODE_TYPE_USER: 'user',
  NODE_TYPE_RESPONSE_GROUP: 'responseGroup',
  NODE_TYPE_NODE: 'node',
  NODE_TYPE_POLICY: 'policy',
  NODE_TYPE_CERTIFICATION: 'certification',
  NODE_TYPE_CLOUD: 'cloud',
  NODE_TYPE_TRACK: 'track',
  NODE_TYPE_SHADOWNODE: 'shadowNode',
};
const NODE_TYPE = undefined;

class AbstractNode extends DataManager {
	constructor() {
		super(...arguments);
		this.nodeType = NODE_TYPE;
	}

	getByAuthorId(authorId) {
		return this.filter(
			node =>
				node.data.authorId === authorId
		);
	}

	/**
	 * Returns the node at the start of this track
	 **/
	getTrailhead(childNode) {
		const parentNode = this.getParentOf(childNode);

		return parentNode === undefined
			? childNode
			: this.getTrailhead(parentNode)
	}

	___move(node, newParentNode) {
		return super.move({ data: node.data.text }, newParentNode);
	}

	_createNodeData(obj) {
		return obj !== undefined
			? super._createNodeData(obj) : false;
	}
};

export {
  NODE_TYPES,
  AbstractNode as default
};