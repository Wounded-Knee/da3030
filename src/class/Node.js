import DataManager from './DataManager';
const NODE_TYPES = {
  NODE_TYPE_USER: 'user',
  NODE_TYPE_RESPONSE_GROUP: 'responseGroup',
  NODE_TYPE_NODE: 'node',
  NODE_TYPE_POLICY: 'policy',
  NODE_TYPE_CERTIFICATION: 'certification',
  NODE_TYPE_CLOUD: 'cloud',
  NODE_TYPE_TRACK: 'track',
};

class Node extends DataManager {
	constructor() {
		super(...arguments);
		this.nodeType = NODE_TYPES.NODE_TYPE_NODE;
	}

	getByAuthorId(authorId) {
		return this.filter(
			node =>
				node.data.authorId === authorId
		);
	}

	/**
	 * Returns the node at the start of this track
	 **/
	getTrailhead(childNode) {
		const parentNode = this.getParentOf(childNode);

		return parentNode === undefined
			? childNode
			: this.getTrailhead(parentNode)
	}

	move(node, newParentNode) {
		return super.move({ data: node.data.text }, newParentNode);
	}

	_createNodeData(text) {
		return text !== undefined
			? super._createNodeData({
				text,
			}) : false;
	}
};

export {
  NODE_TYPES,
  Node as default
};