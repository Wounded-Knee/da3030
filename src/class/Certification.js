import DataManager from './DataManager';
import { NODE_TYPES } from './AbstractNode';

class Certification extends DataManager {
	constructor() {
		super(...arguments);
		this.nodeType = NODE_TYPES.NODE_TYPE_CERTIFICATION;
	}
};

export default Certification;