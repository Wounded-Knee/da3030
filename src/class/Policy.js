import DataManager from './DataManager';
import { NODE_TYPES } from './AbstractNode';

class Policy extends DataManager {
	constructor() {
		super(...arguments);
		this.nodeType = NODE_TYPES.NODE_TYPE_POLICY;
	}
};

export default Policy;