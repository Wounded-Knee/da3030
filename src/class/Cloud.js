import DataManager from './DataManager';
import { NODE_TYPES } from './Node';

class Cloud extends DataManager {
	constructor() {
		super(...arguments);
		this.nodeType = NODE_TYPES.NODE_TYPE_CLOUD;
	}

	getByContributorId(contributorId) {
		return this.filter(
			cloud =>
				cloud.data.contributorId === contributorId
		);
	}

	// Returns info on how this user relates to this cloud.
	// Are they subscribed?
	getUserDisposition(user, cloud) {

	}

	getByUserEligibility(user) {
		return this.filter(
			cloud => {
				return cloud.data.external.qualification.filter(
					nodeId =>
						this.annuitCœptis.Track.userHasTrack(nodeId, user)
				).length === cloud.data.external.qualification.length
			}
		);
	}

	getNotifications(user) {
		return this.getByUserEligibility(user).length;
	}

	_createNodeData(nodeData) {
		const { userId } = this.annuitCœptis.getSettings();

		if (
			userId === undefined ||
			nodeData === undefined
		) return false;

		return super._createNodeData({
			...nodeData,
			contributorId: userId,
		});
	}
};

export default Cloud;