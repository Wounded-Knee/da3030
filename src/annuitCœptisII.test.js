import React from 'react';
import ReactDOM from 'react-dom';
import * as Models from './class/Models';
import AnnuitCÅ“ptisII from './class/AnnuitCÅ“ptisII';
const {
	MODEL_TYPES,
	ATTRIBUTE_NAMES,
	EVENT_TYPES,
} = Models;
const annuitCÅ“ptisII = new AnnuitCÅ“ptisII();
const saveFileName = 'SH10151';
const testTextValue = '1GCEG15W4Y1142815';
const getData = (modelType, iterator) => {
	var data = {
		_meta: {
			id: iterator
		}
	};
	switch (modelType) {
		case MODEL_TYPES.TEXT_NODE:
			return {
				...data,
				text: `This is ${modelType} #${iterator}`
			};
		break;
		case MODEL_TYPES.CLOWN:
			return {
				...data,
				text: `ðŸ¤¡ ${iterator}`
			};
		break;
		default:
			return data;
		break;
	}
};

const testCreateBatchByType = (modelType, qty, getDataCallback = getData) => {
	for (var x=0; x<qty; x++) {
		annuitCÅ“ptisII.create(getDataCallback(modelType, x), modelType);
	}

	expect(
		annuitCÅ“ptisII.filter(
			node => node.getModelType() === modelType
		).length
	).toEqual(qty);
};

it(`Is not dirty`, () => {
	expect(annuitCÅ“ptisII.isDirty()).toEqual(false);
});

it(`Can create 5c GENERIC nodes`, () => {
	testCreateBatchByType(
		MODEL_TYPES.GENERIC,
		500,
	);
});

it(`Can create 5c CLOWN nodes`, () => {
	testCreateBatchByType(
		MODEL_TYPES.CLOWN,
		500,
	);
});

it(`Is dirty due to creating nodes`, () => {
	expect(annuitCÅ“ptisII.isDirty()).toEqual(true);
});

it(`Cannot be clear()'ed`, () => {
	expect(annuitCÅ“ptisII.clear()).toEqual(false);
	expect(annuitCÅ“ptisII.getData().length).toMatchSnapshot();
});

it(`Can save()`, () => {
	annuitCÅ“ptisII.save(saveFileName);
});

it(`Is not dirty`, () => {
	expect(annuitCÅ“ptisII.dirty).toEqual(false);
});

it(`Can be clear()'ed`, () => {
	expect(annuitCÅ“ptisII.clear()).toEqual(true);
	expect(annuitCÅ“ptisII.getData().length).toEqual(0);	
});

it(`Can load()`, () => {
	annuitCÅ“ptisII.load(saveFileName);
	expect(annuitCÅ“ptisII.getData().length).toMatchSnapshot();	
});

it(`Can retrieve an arbitrary node by text value`, () => {
	annuitCÅ“ptisII.clear(true);
	expect(annuitCÅ“ptisII.getData().length).toEqual(0);
	expect(annuitCÅ“ptisII.iterator).toMatchSnapshot();

	testCreateBatchByType(
		MODEL_TYPES.TEXT_NODE,
		3,
		(modelType, iterator) => ({
			...getData(modelType, iterator),
			text: testTextValue,
		}),
	);

	const results = annuitCÅ“ptisII.filter(
		node => node.get('text') === testTextValue
	);
	expect(results.length).toEqual(3);
	expect(results[2].getId()).toMatchSnapshot();
});

it(`Can have children`, () => {
	const eventType = EVENT_TYPES.ADOPTEDBY;
	annuitCÅ“ptisII.clear(true);

	testCreateBatchByType(
		MODEL_TYPES.GENERIC,
		1914,
	);

	const [
		parent,
		child
	] = annuitCÅ“ptisII.getByModelType(MODEL_TYPES.GENERIC);
	expect(parent.represent()).toMatchSnapshot();
	expect(child.represent()).toMatchSnapshot();
	expect(child.getEventsByType(eventType).length).toEqual(0);
	child.setParent(parent);
	expect(child.getEventsByType(eventType).length).toEqual(1);
	expect(parent.getChildren()[0].getId()).toEqual(child.getId());
	expect(child.getParent().getId()).toEqual(parent.getId());
});

it(`Can be orphaned`, () => {
	const [
		,child
	] = annuitCÅ“ptisII.getByModelType(MODEL_TYPES.GENERIC);
	expect(child.getParent()).not.toBeFalsy();
	child.orphan();
	expect(child.getParent()).toBeFalsy();
});

it(`Can set data by attribute`, () => {
	const testTextValue2 = 'H.Con.Res.331-1988';
	testCreateBatchByType(
		MODEL_TYPES.TEXT_NODE,
		3,
		(modelType, iterator) => ({
			...getData(modelType, iterator),
			text: testTextValue2,
		}),
	);

	const [ node ] = annuitCÅ“ptisII.getByModelType(MODEL_TYPES.TEXT_NODE);
	expect(node.get('text')).toMatchSnapshot();
	node.set(testTextValue, 'text');
	expect(node.get('text')).toMatchSnapshot();
});

it(`Records audit trail in events`, () => {
	const [ node ] = annuitCÅ“ptisII.getByModelType(MODEL_TYPES.TEXT_NODE);
	node.set(testTextValue, 'text');
	const events = node.getEventsByType(EVENT_TYPES.SET_DATA);
	expect(events[0].type).toMatchSnapshot();
	expect(events.length).toMatchSnapshot();
});
