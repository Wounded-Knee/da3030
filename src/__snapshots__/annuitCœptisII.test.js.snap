// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Can have children 1`] = `"Generic 1003"`;

exports[`Can have children 2`] = `"Generic 1004"`;

exports[`Can load() 1`] = `1000`;

exports[`Can retrieve an arbitrary node by text value 1`] = `1000`;

exports[`Can retrieve an arbitrary node by text value 2`] = `1002`;

exports[`Can set data by attribute 1`] = `"H.Con.Res.331-1988"`;

exports[`Can set data by attribute 2`] = `"1GCEG15W4Y1142815"`;

exports[`Cannot be clear()'ed 1`] = `1000`;

exports[`Records audit trail in events 1`] = `"Set Data by Attribute"`;

exports[`Records audit trail in events 2`] = `2`;
